### ISSUE TITLE
Reverse mounting doesn't work in Mac OS X?

### ISSUE THREAD

**Issue by Nitrodist** [dustymabe/vagrant-sshfs#46]
Hey, I'm a "1%" user who wants to mount a directory from the guest OS onto my host machine so I can use a text editor on the host OS to edit the files that exist on the machine.

I'm running El Capitan (10.11.6) with Vagrant 1.8.5, and VirtualBox 5.0.20. 

The specific error I get is this:

```
==> default: Mounting NFS shared folders...
The capability 'sshfs_reverse_is_folder_mounted' could not be found. This is an internal error
that users should never see. Please report a bug.
```

The line in my Vagrant file is this:

``` diff
diff --git a/Vagrantfile b/Vagrantfile
index b533225..7ca217b 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -30,4 +30,6 @@ Vagrant.configure('2') do |config|
     vbox.customize(['modifyvm', :id, '--natdnshostresolver1', 'on'])
     vbox.customize(['modifyvm', :id, '--natdnsproxy1', 'on'])
   end
+
+  config.vm.synced_folder "/Users/mark.campbell/Desktop/vagrant/", "/home/vagrant/src", type: "sshfs", sshfs_opts_append: "-o cache=no", reverse: true
 end
```

I poked around in the code and I noticed that the `host_capability` line is this

``` ruby
      host_capability("linux", "sshfs_reverse_is_folder_mounted") do
        require_relative "cap/host/linux/sshfs_reverse_mount"
        VagrantPlugins::HostLinux::Cap::MountSSHFS
      end
```

I don't know much, but I don't think that the "linux" value will let my host machine which is mac os x use this feature. 

Is this a matter of updating the documentation as such or can we add this a feature just for mac os x? Or should this feature actually be working under mac os x?

**Comment by dustymabe** [comment]
Hi @Nitrodist. TBH I didn't have a mac to test this on so I didn't add support for MAC initially. I assume you have installed some version of sshfs on your mac. Is that correct? Can you link me to the version you installed?

**Comment by Nitrodist** [comment]
```
$ sshfs --version
SSHFS version 2.8
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.3
no mount point
```

**Comment by dustymabe** [comment]
@Nitrodist thanks - I'll try to grab a mac system and pull this in to the next release. It shouldn't be hard to do at all.

**Comment by dantheman213** [comment]
Did this ever get resolved? I'm also experiencing a similar error.

**Comment by dustymabe** [comment]
@dantheman213 - not yet, still need to get my hands on a mac. I had access to one in the past but not now. Will try to get around to this soon.

**Comment by dantheman213** [comment]
@dustymabe they're relatively easy to virtualize nowadays. Check this out? https://techsviewer.com/how-to-install-mac-os-x-el-capitan-on-vmware-on-pc/

**Comment by neerolyte** [comment]
I'd like to be able to do this to.

With https://github.com/osxfuse/osxfuse/wiki/SSHFS + https://osxfuse.github.io/ I'm able to manually do:
```
$ vmname=foo
$ mkdir sshfs
$ vagrant ssh-config "$vmname" > ."$vmname".sshconfig
$ sshfs -F "$PWD"/."$vmname".sshconfig "$vmname":/var/www sshfs -o sftp_server="sudo /usr/libexec/openssh/sftp-server",auto_cache,reconnect,defer_permissions,noappledouble
```

and unmount with:

```
umount sshfs
```

I think aside from some of the `-o` options that should be pretty similar to linux?

**Comment by bcah** [comment]
@dustymabe, we are also part of the 1%, this functionality would help us immensely.

**Comment by dantheman213** [comment]
https://media.giphy.com/media/13pge2TFoz4kN2/giphy.gif

**Comment by dustymabe** [comment]
hey everyone can you try the following and tell me if your problem is fixed? 

- download the file at https://dustymabe.fedorapeople.org/vagrant-sshfs-1.2.1.a7d8a89.gem
- `vagrant plugin uninstall vagrant-sshfs`
- `vagrant plugin install ./vagrant-sshfs-1.2.1.a7d8a89.gem`
- test to see if your problem is fixed

**Comment by dantheman213** [comment]
Just to be clear, this will enable my vagrant install to mount a folder between host/guest while preserving the files/folders in the guest at the target mount path? 

I went through a significant workaround for my vagrant box to place all my server configs on the host machine and mount them down to the guest.

**Comment by dustymabe** [comment]
> Just to be clear, this will enable my vagrant install to mount a folder between host/guest while preserving the files/folders in the guest at the target mount path?

Yes. The files live on the guest and get mounted into the host so they can be accessible there. If you `vagrant destroy` the VM, the files will be destroyed as well.

**Comment by dustymabe** [comment]
hey everyone, here is a new dev version to try. The other one had a bug in it (unrelated to this issue):


- download the file at https://dustymabe.fedorapeople.org/vagrant-sshfs-1.2.1.b3eae35.gem
- `vagrant plugin uninstall vagrant-sshfs`
- `vagrant plugin install ./vagrant-sshfs-1.2.1.b3eae35.gem`
- test to see if your problem is fixed

**Comment by dantheman213** [comment]
I'll check this out tomorrow. Thanks again!

**Comment by lanej** [comment]
Working for me @dustymabe. Thanks!

**Comment by neerolyte** [comment]
It mostly works, but I think without [defer_permissions](https://github.com/osxfuse/osxfuse/wiki/Mount-options#default_permissions-and-defer_permissions) set most users won't be able to edit most files - i.e if the uids don't match between the owner in the VM and the current user in OS X (and they don't do anything special to map them) they won't be able to edit them.

tldr - it only works if I add:

```
sshfs_opts_append: "-o defer_permissions"
```

to the synced_folder line on OS X, which means having a different line for OS X v.s Linux users.

**Comment by dustymabe** [comment]
defer_permissions doesn't seem to be in the linux version of sshfs. This looks similar to https://github.com/dustymabe/vagrant-sshfs/issues/20. Either way, you have a workaround for now.

**Comment by vincenzo** [comment]
@dustymabe Are you going to make an official release with the bug fix contained in `vagrant-sshfs-1.2.1.b3eae35.gem`?

**Comment by vincenzo** [comment]
I am not using `defer_permissions`, btw (I am on Mac), but these: 

```
mount_options: ['allow_other', 'workaround=all', 'nonempty', 'reconnect']
```

**Comment by dustymabe** [comment]
> @dustymabe Are you going to make an official release with the bug fix contained in vagrant-sshfs-1.2.1.b3eae35.gem?

@vincenzo, yes. I plan to do that soon

**Comment by vincenzo** [comment]
Cheers, @dustymabe.

**Comment by jjn1056** [comment]
similar issue for me... I can get my folders mounted (after rewriting all the config files) but the permissions are wrong and can't seem to get it working no matter how I mess around with the sshfs options.  There's other problems as well, like the mount point doesn't always get unmounted when I halt the VM.

**Comment by fedealconada** [comment]
@dustymabe has this issue been fixed? I cannot make it work... I'm being able to mount, but if I run `ls -l` inside the mounted folder I get:
`ls: .: Device not configured`
Any ideas?

**Comment by dustymabe** [comment]
> @dustymabe has this issue been fixed? I cannot make it work... I'm being able to mount, but if I run ls -l inside the mounted folder I get:
> ls: .: Device not configured
> Any ideas?

@fedealconada, should be fixed. can you open a new issue with as much detail as possible for the issue you are seeing?

> 
> P.S Oh, BTW, I had to comment out line 74 (ssh_opts+= ' -o "IdentityFile=\"' + machine.ssh_info[:private_key_path][0] + '\""') from sshfs_reverse_mount.rb file because it was giving some problems...

A separate issue with more details on this error would be useful as well.
